packages/slice-machine/src/features/onboarding/OnboardingProvider.tsx

```diff
@@ -91,5 +104,9 @@ export const useOnboardingContext = () => {
     );
   }
 
+  if (onComplete) {
+    context.setOnCompleteCallback(onComplete);
```

Comment:
There are some problems with this approach. We're performing a side effect on render, and also different useOnboardingContext usages could override the onComplete callback from others.

A better approach is to move this to the provider directly (pass `onComplete` to `OnboardingProvider`), this is easy to do if you move the conffeti implementation to `OnboardingGuide`.

Reply:
Does it matter if the only thing we're doing it assigning a callback to a ref?

Reply:
but the other problem is also important, imagine having two components setting the callback in every render, overriding one another

Reply:
Man, react and it's never ending pitfalls. They say it makes the `"component’s behavior difficult to predict"` but I don't even know how or why 😅

Reply:
react is nice but it's definitely not perfect

i asked dan abramov in the past what that "unpredictable behavior" means in practice but he never replied

i understand it has to do with the concurrent mode and the ability for the components to pause/cancel rendering and start back again later

i also agree with you that the rules of hooks are annoying and that they shouldn't have been just eslint rules, as they're too important; hopefully, that'll change with the new react compiler (which is also a reason to respect the rules now, so we get an easier path to migrate to it in the future)

i think the best we can do is to learn these rules and caveats together and help each other find ways of implementing things the "react way", even if sometimes we get annoyed by it

in the end, react adds a lot of value and it's a nice way of building UIs

Reply:
Changed, I avoided passing the callback directly to the provider because I wanted to container everything in the "Content" component, but I extract an intermediate one and I'm happy with it. Tell me what you think.

Reply:
Yeah I'm used to these rules and caveats, but the ref one I wasn't, to be honest 😄